src/dune_rules/melange/melange_rules.ml

```diff
@@ -52,6 +52,21 @@ let impl_only_modules_defined_in_this_lib sctx lib =
           (Lib.name lib |> Lib_name.to_string)
       ]
   | Some modules ->
+    let () =
```

Comment:
Ended up adding the check at this point to make sure we don't error twice for the case found in #7020, see related commit: https://github.com/ocaml/dune/pull/6412/commits/e02e6ffb330fe4ce3ce34659b06d6656cd40ec19.